server/modules/module_state_machine.js

```diff
@@ -263,16 +269,8 @@ class DisplayState extends stateMachine.State {
     // duration pass.
     let durationBeforePrepare = Math.max(this.displayDuration_ - 5000, 0);
     let prepareDeadline = this.deadline_ + durationBeforePrepare;
-    let fadeStartDeadline = this.deadline_ + this.displayDuration_;
     debug('Waiting until', this.deadline_, prepareDeadline);
-    this.timer_ = setTimeout(() => {
-      if (this.playlist_.length == 0) {
-        throw new Error('No modules left in playlist!');
-      }
-      let nextModuleIndex = (this.index_ + 1) % this.playlist_.length;
-      debug(`Begin preparing to transition to ${this.playlist_[nextModuleIndex].name} (item ${nextModuleIndex} of ${this.playlist_.length})`);
-      this.transition_(new DisplayState(this.playlist_, this.layout_, fadeStartDeadline, nextModuleIndex));
-    }, time.until(prepareDeadline));
+    this.timer_ = setTimeout(() => this.nextModule(), time.until(prepareDeadline));
```

Comment:
Should you restart the timer if you're told to change modules?  Or on an error?

Reply:
We actually DO restart the timer in these cases, but it's super circuitous (which is part of the motivation for moving this playlist management stuff out of the SMs). In the case of an error, we eventually end up transitioning to the next module in the playlist, which eventually calls this line again, restarting the timer. Same logic around the next module call.